sql/sql_table.cc

```diff
@@ -12085,9 +12085,17 @@ copy_data_between_tables(THD *thd, TABLE *from, TABLE *to, bool ignore,
     error= 1;
   }
   bulk_insert_started= 0;
-  if (!ignore)
-    to->file->extra(HA_EXTRA_END_ALTER_COPY);
-
+  if (!ignore && error <= 0)
+  {
+    int alt_error= to->file->extra(HA_EXTRA_END_ALTER_COPY);
+    if (alt_error == HA_ERR_FOUND_DUPP_KEY)
+    {
+      uint key_nr= to->file->get_dup_key(alt_error);
+      if ((int)key_nr >= 0 && key_nr < to->s->keys)
+        print_keydup_error(to, &to->key_info[key_nr], MYF(0));
+    }
+    error= alt_error;
+  }
```

Comment:
Again, why not `!error` instead of `error <= 0`? In this case, we could overwrite a negative `error` code with `alt_error=0`.

Reply:
error is set to -1 when handler encounters HA_ERR_END_OF_FILE. So error variable is > 0 then we encounter error.